### ISSUE TITLE
Inconsistent handling of dates before 01.01.1900

### ISSUE THREAD

**Issue by oliver-huetkoeper** [abap2xlsx/abap2xlsx#1275]
Hi,

currently absp2xlsx handles „invalid“ Excel dates (dates before 01.01.1900) different on the import and on the export side. 
I want to convert an ALV to a CSV file using abap2xlsx. If the data contains such a date, converting the ALV to Excel works without problem. But exporting it to CSV fails. 
In my opinion, this is inconsistent. What do you think? 

Best regards,
Oliver

```abap
PARAMETERS p_path TYPE string LOWER CASE.

INITIALIZATION.
  cl_gui_frontend_services=>get_sapgui_workdir( CHANGING sapworkdir = p_path ).
  cl_gui_cfw=>flush( ).

START-OF-SELECTION.
  TYPES tt_sflight TYPE STANDARD TABLE OF sflight WITH EMPTY KEY.

  DATA(gt_sflight) = VALUE tt_sflight(
      ( carrid = 'AA' connid = 1 fldate = '20000101'  )
      ( carrid = 'AA' connid = 1 fldate = '18000101'  ) ).
  cl_salv_table=>factory( IMPORTING r_salv_table = DATA(lo_salv)
                          CHANGING  t_table      = gt_sflight ).
  cl_salv_bs_runtime_info=>set( display  = abap_false
                                metadata = abap_false
                                data     = abap_false ).
  lo_salv->display( ).

  DATA(lo_converter) = NEW zcl_excel_converter( ).
  DATA(lo_excel) = NEW zcl_excel( ).
  lo_converter->convert( EXPORTING io_alv       = lo_salv
                                   it_table     = gt_sflight
                         CHANGING  co_excel     = lo_excel ).

  DATA(lo_excel_writer) = CAST zif_excel_writer( NEW zcl_excel_writer_csv( ) ).
  DATA(lv_file) = lo_excel_writer->write_file( lo_excel ).

  DATA(lt_file_tab) = cl_bcs_convert=>xstring_to_solix( lv_file ).
  cl_gui_frontend_services=>gui_download( EXPORTING bin_filesize = xstrlen( lv_file )
                                                    filename     = p_path && '\Sheet1.csv'
                                                    filetype     = 'BIN'
                                          CHANGING  data_tab     = lt_file_tab ).
```

**Comment by gregorwolf** [comment]
Do you have a screenshot with the differences?

**Comment by oliver-huetkoeper** [comment]
> Do you have a screenshot with the differences?

The converter creates negative numbers from such a date. The CSV writer raises an exception if the number is negative.
Method date_to_excel_string vs. excel_string_to_date in class zcl_excel_common.

**Comment by sandraros** [comment]
Result of the program above ([Sheet1.csv](https://github.com/user-attachments/files/17210480/Sheet1.csv)):
![image](https://github.com/user-attachments/assets/273c1dcc-16c4-413a-a48e-0677367857fe)

(36526 is the Excel date number representing the YYYYMMDD date `20000101`, -36523 is for `18000101`)

**Comment by sandraros** [comment]
Obviously, a bug to be fixed. It must have been here for a long time; my 2 cents that it originates from ancient modifications of `zcl_excel_worksheet=>set_cell` concerning "data type" and "value type" notions (which I never clearly understood). In the program above, `zcl_excel_writer_csv` receives an empty data type for date and number fields, but the class expects "D" for date fields (there's also a crazy logic which tests the field names to determine the data type).

**Comment by oliver-huetkoeper** [comment]
It looks like the texts of domain XUDATFM changed in newer systems:
NW 7.40: DD.MM.YYYY
NW 7.52: DD.MM.YYYY (Gregorian Date).

This breaks the crazy date identification logic.
I implemented a fix, which is part of https://github.com/abap2xlsx/abap2xlsx/pull/1268

Now the demo program also dumps in newer systems ;)
So what is your opinion on the invalid date handling on import/export side?

**Comment by darnoc312** [comment]
1) I think @oliver-huetkoeper was using 'E' as system-language and @sandraros did not.
That is why @oliver-huetkoeper got an exception and @sandraros got a result with negative values.

There is the following code in method create_csv
```abap
* --- Retrieve SAP date format
  CLEAR ls_format.
  SELECT ddtext INTO ls_format-attvalue FROM dd07t WHERE domname    = 'XUDATFM'
                                                     AND ddlanguage = sy-langu.     "THIS IS WRONG
    ls_format-cmpname = 'DATE'.
    CONDENSE ls_format-attvalue.
    CONCATENATE '''' ls_format-attvalue '''' INTO ls_format-attvalue.
    APPEND ls_format TO lt_format.
  ENDSELECT.
```

Please refer to method zcl_excel_worksheet->GET_DEFAULT_EXCEL_DATE_FORMAT:
There you find the following code:
```abap
      cl_abap_datfm=>get_date_format_des( EXPORTING im_langu = c_lang_e
                                          IMPORTING ex_dateformat = default_excel_date_format ).
```

The date formats like 'DD.MM.YYYY' and similar corresponds to english language only and should be read in english above, too.

2) After this fix we get always the exception. With issue #703 some code was added in zcl_excel_common->EXCEL_STRING_TO_DATE. Of course this is correct for normal excel output. But perhaps we should add a flag IP_PERMIT_NEGATIVE to be used from method create_csv only. Then the lower boundary to check is -693596 (01.01.0000)

**Comment by oliver-huetkoeper** [comment]
@darnoc312 you are right regarding the language bug. I also discovered it and there is a fix for this in my pull request as well. 
I was not sure if English is always correct. Therefore I SELECT now with sy-langu and with E.

**Comment by darnoc312** [comment]
In my opinion only language 'E' makes sense, because excel itself can not deal with 'JJJJ' or 'AAAA' as synonym for year in the format code. But you could argue to permit sy-langu because of backwards compatibility for csv only. So far the local date format was detected as date for csv although it does not work for excel. But if you like the possibility to switch the writer in your application program this may be another hidden problem. May @sandraros decide.

**Comment by oliver-huetkoeper** [comment]
> 1. I think @oliver-huetkoeper was using 'E' as system-language and @sandraros did not.
>    That is why @oliver-huetkoeper got an exception and @sandraros got a result with negative values.
> 
> There is the following code in method create_csv
> 
> ```abap
> * --- Retrieve SAP date format
>   CLEAR ls_format.
>   SELECT ddtext INTO ls_format-attvalue FROM dd07t WHERE domname    = 'XUDATFM'
>                                                      AND ddlanguage = sy-langu.     "THIS IS WRONG
>     ls_format-cmpname = 'DATE'.
>     CONDENSE ls_format-attvalue.
>     CONCATENATE '''' ls_format-attvalue '''' INTO ls_format-attvalue.
>     APPEND ls_format TO lt_format.
>   ENDSELECT.
> ```
> 
> Please refer to method zcl_excel_worksheet->GET_DEFAULT_EXCEL_DATE_FORMAT: There you find the following code:
> 
> ```abap
>       cl_abap_datfm=>get_date_format_des( EXPORTING im_langu = c_lang_e
>                                           IMPORTING ex_dateformat = default_excel_date_format ).
> ```
> 
> The date formats like 'DD.MM.YYYY' and similar corresponds to english language only and should be read in english above, too.
> 
> 2. After this fix we get always the exception. With issue [zcl_excel_common=>excel_string_to_date returns wrong value when cell was empty #703](https://github.com/abap2xlsx/abap2xlsx/issues/703) some code was added in zcl_excel_common->EXCEL_STRING_TO_DATE. Of course this is correct for normal excel output. But perhaps we should add a flag IP_PERMIT_NEGATIVE to be used from method create_csv only. Then the lower boundary to check is -693596 (01.01.0000)

Looking at the code introduced with https://github.com/abap2xlsx/abap2xlsx/issues/703 it is basically the check statement at the beginning that solved the issue back then. 
Therefore I would propose to simply extend the valid value range and do not introduce a new parameter. Or do you see any need to restrict this to only valid Excel dates?

**Comment by kartefact** [comment]
The `date_to_excel_string` method correctly handles dates before 1900 by producing negative numbers (like -36523 for 18000101), but `excel_string_to_date` has validation that rejects negative values `zcl_excel_common.clas.abap:895-897` .

### CSV Writer Issue
The CSV writer calls `excel_string_to_date` when processing date cells `zcl_excel_writer_csv.clas.abap:275` , which throws an exception for the negative values that represent pre-1900 dates.

### Proposed Solutions
Based on the discussion in the issue, there are two main approaches:

### Option 1: Add a Parameter to Allow Negative Values
Add an `ip_permit_negative` parameter to `excel_string_to_date` that the CSV writer can use to bypass the validation for negative values. The lower boundary would be -693596 (representing 01.01.0000).

### Option 2: Extend the Valid Range (Recommended)
Simply extend the valid value range in `excel_string_to_date` to accept negative values without requiring a new parameter. This would make the behavior consistent across all writers.

### Implementation
The fix should modify the validation in `excel_string_to_date` to accept a wider range of values. 
Currently it checks:
`IF lv_date_int NOT BETWEEN 1 AND 2958465.`

This should be changed to allow negative values down to approximately -693596 (representing year 0000).

Can anyone confirm if this is correct?
